shopfloor/services/cluster_picking.py

```diff
@@ -490,21 +490,31 @@ def _scan_line_by_product(self, picking, move_line, product):
         """Product scanned, check if we can work with it.
 
         If scanned product is part of several packages in the same location,
-        we can't be sure it's the correct one, in such case, ask to scan a package
+        we can't be sure it's the correct one, in such case, ask to scan a package.
+
+        If the product is tracked by lot and there is only one lot id in the location
+        not in a package. It can safely be picked up.
         """
-        if move_line.product_id.tracking in ("lot", "serial"):
-            return self._response_for_start_line(
-                move_line, message=self.msg_store.scan_lot_on_product_tracked_by_lot()
-            )
-        other_product_lines = picking.move_line_ids.filtered(
-            lambda l: l.product_id == product and l.location_id == move_line.location_id
+        message = None
+        location_quants = move_line.location_id.quant_ids.filtered(
+            lambda quant: quant.quantity > 0 and quant.product_id == product
         )
```

Comment:
A `search` on `stock.quant` will be better I guess here, if you get thousands of quants in your location, it'll loop over all of them with `filtered`.

Also I'm wondering: we will grab quants that are not related to the current `picking`, even some reserved by other operations, is it what we want?! The packages and lots we are looking for should be those already reserved in `picking` IMO. cc @jbaudoux

Reply:
This commit and the [next one](https://github.com/OCA/wms/pull/443/commits/99a7cbdc2802f141fb285d5a10815a6f2aa1f55f) are titled

```
shopfloor: fix cluster picking _scan_line_by_product

Check the content of the location not the move lines of the picking.
```
```
shopfloor: fix cluster picking _scan_line_location

Fix the error raised if there are multiple packs, products or lot in the location.
Not based on move lines but based on actual stock at location.
```
So it has been done on purpose, but I do not understand why :thinking:

Reply:
@sebalix You are in front of a location and we want to be sure you take the right product from that location. So it must look at what is on the location, not what is in the processed picking.
I think filtered is fine, there won't be much quants on a location bin.

Reply:
Same for lot. If there are multiple lots on the location, you have to scan the right one. Scanning the product is only enough if there is only a single lot on the location. No matter what are the picking move lines

Reply:
This implies to use location bin, with a too big source location we'll fetch too many things. Perhaps it'll never happen, I haven't the big picture, because it's like we developed this feature to fit our customers config, while it could not work well in another context. I could be wrong.